### ISSUE TITLE
Support splitting into multiple objects?

### ISSUE THREAD

**Issue by fzyzcjy** [Voultapher/self_cell#58]
Hi thanks for the helpful crate!

When implementing https://github.com/fzyzcjy/flutter_rust_bridge, I am considering the following scenario. Firstly a simple one:

```rust
struct One { .. }
struct Two<'a> { .. }
fn compute(one: &'a One) -> Two<'a> { .. }
```

I have a `Arc<One>` as input, and want to provide something as if `Arc<Two>` as output, but it should have no lifetime specifier by putting `Arc<One>` inside it.

```rust
// pseudo code for both self_cell and ouroboros
struct TwoWithOwner {
  one: Arc<One>,
  two: Two<'this>,
}

let input: Arc<One> = ...;
let output: Arc<TwoWithOwner> = Arc::new(TwoWithOwnerBuilder {
  one: input,
  two_builder: |one| compute(one),
}.build());
```

Now comes the problem. Suppose the return type is a `Vec`:

```rust
fn compute(one: &'a One) -> Vec<Two<'a>> { .. }
```

And for the same input, I hope to provide `Vec<Arc<Two>>` as output (again without lifetimes).
If using the exact same code, I will only get something like:

```rust
struct VecTwoWithOwner {
  one: Arc<One>,
  two: Vec<Two<'this>>,
}
```

In other words, an `Arc<VecTwoWithOwner>` object, instead of `Vec<Arc<TwoWithOwner>>`.

It would be great if there are some API that can split an `Arc<VecTwoWithOwner>` object into multiple `Vec<Arc<TwoWithOwner>>` objects. From my naive thoughts, maybe there can be something like a `transform` function. Or, there may be better solutions for the scenario above.

Cross-posted in: https://github.com/someguynamedjosh/ouroboros/issues/117

**Comment by Voultapher** [comment]
With this as basis for your question:

```rust
struct One(String);
struct Two<'a>(&'a One);

fn compute<'a>(one: &'a One) -> Vec<Two<'a>> {
    vec![Two(one), Two(one)]
}

self_cell!(
    struct SelfCell {
        owner: One,

        #[covariant]
        dependent: Two,
    }
);

let cell = SelfCell::new(One("xx".into()), compute);
```

I think there is no general solution that would allow splitting without making assumptions about the owner type, such as address stability. This has been discussed in depth here https://github.com/Voultapher/self_cell/issues/55.

However, maybe there is a way forward for your issue. Given the above example code, assuming you have control over `compute` you could change it to `fn compute(one: Arc<One>) -> Vec<SelfCell>`, constructing one cell per dependent type. If `compute` can't be changed, I would look at the higher level view, why is there are 1-N mapping from `One` to `Two` but bundling them in a way where each `Two` comes with its own `One` makes sense. Maybe it's possible to hoist the structure and wrap it in an `Arc`, so that you have:


```rust
use std::sync::Arc;

struct One(String);
struct Two<'a>(&'a One);

fn compute<'a>(one: &'a One) -> Vec<Two<'a>> {
    vec![Two(one), Two(one)]
}

type Dependent<'a> = Vec<Two<'a>>;

self_cell!(
    struct SelfCell {
        owner: One,

        #[covariant]
        dependent: Dependent,
    }
);

let cell = Arc::new(SelfCell::new(One("xx".into()), compute));
let cell_cheap_clone = cell.clone();
let two_idx_1 = &cell_cheap_clone.borrow_dependent()[1];
```

Does that help?

**Comment by fzyzcjy** [comment]
Hi, thank you for the reply!

> I think there is no general solution that would allow splitting without making assumptions about the owner type, such as address stability. This has been discussed in depth here https://github.com/Voultapher/self_cell/issues/55.

I agree. Thus is it possible to deal with it, when the owner has type `Arc<Something>`?

>  assuming you have control over compute you could change it to fn compute(one: Arc<One>) -> Vec<SelfCell>

This is user code, thus I cannot control it...

> If compute can't be changed, I would look at the higher level view, why is there are 1-N mapping from One to Two but bundling them in a way where each Two comes with its own One makes sense. Maybe it's possible to hoist the structure and wrap it in an Arc, so that you have:

Thank you, but this may not be ideal, because I need to convert the `Vec<SelfCell>` into a real Dart list.

**Comment by Voultapher** [comment]
> I agree. Thus is it possible to deal with it, when the owner has type Arc<Something>?

As explained in the linked issue, this is a generic library and it doesn't know what `Arc` like types are, and even then we would have to limit the borrow to interior things.


If all you need is a self-contained thing you can easily store and pass around, would the outlined `Arc<SelfCell>` approach work, where you assemble the interleaved Dart object as needed on the FFI boundary?

**Comment by fzyzcjy** [comment]
> As explained in the linked issue, this is a generic library and it doesn't know what Arc like types are, and even then we would have to limit the borrow to interior things.

Thank you. I was thinking about something like "specializing" on Arc-like things.

> If all you need is a self-contained thing you can easily store and pass around, would the outlined Arc<SelfCell> approach work, where you assemble the interleaved Dart object as needed on the FFI boundary?

I can make the vec an "opaque" object on Dart side, but that may be less than ideal, since users may want a real dart list, and other parts of flutter_rust_bridge also gets better when it is a real list.

**Comment by Voultapher** [comment]
> Thank you. I was thinking about something like "specializing" on Arc-like things.

Please read the linked issue, it goes into depth why that's not an option for me.

> I can make the vec an "opaque" object on Dart side, but that may be less than ideal, since users may want a real dart list, and other parts of flutter_rust_bridge also gets better when it is a real list.

I have no experience with Dart, my idea was to construct the lifetime dependent `Vec<(One, Two<'a>)>` when needed, so as to have a container that is easy to store and pass around in the self_cell and only having to deal with lifetime in the usage place.

**Comment by Voultapher** [comment]
Given that I don't plan on going down the path of specializing on Arc-like constructs, a pre-requisite for the feature in question, I'm proposing to close this issue.

**Comment by fzyzcjy** [comment]
It's OK, thank you all the same!

> to construct the lifetime dependent Vec<(One, Two<'a>)> when needed,

Not sure whether I get the idea, but suppose the `One` is very heavy to be created, and `Two<'a>` is also heavy to be computed from the `One` instance, then it seems not possible to construct one-the-fly.

**Comment by Voultapher** [comment]
> Not sure whether I get the idea, but suppose the One is very heavy to be created, and Two<'a> is also heavy to be computed from the One instance, then it seems not possible to construct one-the-fly.

The idea is to only compute them once, and store them in:

```rust
type Dependent<'a> = Vec<Two<'a>>;

self_cell!(
    struct SelfCell {
        owner: One,

        #[covariant]
        dependent: Dependent,
    }
);
```

The layout is not right yet for the final use, if I understand correctly you need to interleave them into `Vec<(One, Two<'a>)>` to map to the Dart list representation. All that would need to be assembled on-the-fly would be the final `Vec`, which is relatively cheap I guess.

**Comment by fzyzcjy** [comment]
Well I still do not quite get it. What should I fill in `???` below:

```rust
let intermediate_output: SelfCell = ...; // (One, Vec<Two<'a>>)
let real_output: Vec<SelfCell2> = _____???______;

self_cell!(
    struct SelfCell2 {
        owner: Arc<One>, 
        #[covariant]
        dependent: Two,
    }
);
```

**Comment by Voultapher** [comment]
Why do you need `SelfCell2` at all? People usually erase lifetimes to make storage simpler. With `SelfCell` you already have all the information computed once, and in one easy to pass around package.

**Comment by Voultapher** [comment]
It would help if you'd explain what the usage of the desired data structure actually looks like.

**Comment by fzyzcjy** [comment]
Thank you!

In short, since we want a real Dart list (that can, for example, interleave some Rust objects with some Dart objects), we have to have a `Vec<Something>`. Then, suppose it is indeed `[a,b,c]` three objects, then it will be converted to a Dart list (the list is on Dart heap) with three Rust objects (the objects are on Rust memory).

Btw there are some workarounds, but they will make flutter_rust_bridge more complicated IMHO, so I do not elaborate them here.

**Comment by Voultapher** [comment]
I'm more of a visual learner, could you maybe provide a slimmed-down but representative sample program that shows the user-provided `compute` function and then how that Dart list should be used. Right now I'm struggling to wrap my head around it.

**Comment by fzyzcjy** [comment]
Hi, I have done implementing the feature without self_cell, but still thank you for willing to help! I may try to find some time drawing a diagram, but there are still many things to do in frb thus quite busy now. Thus maybe close this issue as you have suggested above firstly.
